core/state/snapshot/conversion.go

```diff
@@ -88,6 +90,54 @@ func VerifyState(snaptree *Tree, root common.Hash) error {
 	return nil
 }
 
+// CommitAndVerifyState takes the whole snapshot tree as the input, traverses all the
+// accounts as well as the corresponding storages and commits all re-constructed trie
+// nodes to the given database(usually the given db is not the main one used in the
+// system, acts as the temporary storage here).
+//
+// Besides, whenever we meet an account with additional contract code, the code will
+// also be migrated to ensure the integrity of the newly created state.
+func CommitAndVerifyState(snaptree *Tree, root common.Hash, db ethdb.Database, commitdb ethdb.KeyValueWriter) error {
```

Comment:
I'm unsure about `db ethdb.Database`. Whenever we access trie data, it should really be done via `trie.Database`, since during runtime that can contain temporary node that have not yet reached the disk layer. I do understand that in this very specific use case operates offline while the node is not running. Perhaps in that case we should explicitly mention that limitation in the method docs. It's an important detail as if you call the method on a running node, it will blow up.

Nit suggestions:

- Rename the method to GenerateTrie. I think it's more suitable and obvious what it does. *Commit* to me says that take some existing data and just push it to disk, whereas this method's main goal is to generate data.
- Rename `db` to `src` and `commitdb` to `dst`. Imho it's clearer that way.
- I'd move this method and `VerifyState` too to be a method on `snaptree *Tree`. There's no point to have package-dangling methods that take their first parameter the main struct of the package.

Reply:
> For the ethdb.Database

Actually it's not. The `src` database is only for reading the snapshot and contract code. There is no extra caching for the snapshot data. And also previously it's true that the contract code may be cached in the `trie.Database` while after changing the contract code scheme they are committed directly. So they should all be available in the disk.